### PR TITLE
Add functions Array::find_first() and Array::get_first()

### LINKED ISSUE #6889: Add more array methods: first_custom, last_custom, count_custom
### Describe the project you are working on

This is not unique to my project, but one project I am working on requires a lot of array traversals searching for elements that satisfy specific conditions. I feel there is a gap in Array functionality (because sort_custom already exists), and this would plug it,

### Describe the problem or limitation you are having in your project

There is no built-in methods to search for the first, last, and to count elements that satisfy custom conditions implemented in a Callable, the way you can filter or sort_custom.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add the following methods:
`find_custom(fn:Callable)` will return the index of the first element of an array where the callable fn will return true.
`rfind_custom(fn:Callable)` will return the index of the last element of an array where the callable fn will return true.
`count_custom(fn:Callable)` will return the count of all elements of the array where the callable fn will return true.

I attempted to name the functions to fit the naming style of the rest of the functions in the Array class. However, I do not like it- I would rather name it something like `find_by()`. But my gripes with the Array function naming conventions are beyond the scope of this proposal.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Here is a sketch implementation I have already created:
```cpp
int Array::find_custom(const Callable& p_callable, int p_from = 0) const {
	if (_p->array.size() == 0) {
		return -1;
	}

	if (p_from < 0 || size() == 0) {
		return -1;
	}

	const Variant* argptrs[1];

	for (int i = p_from; i < size(); i++) {
		argptrs[0] = &get(i);

		Variant result;
		Callable::CallError ce;

		p_callable.callp(argptrs, 1, result, ce);
		if (ce.error != Callable::CallError::CALL_OK) {
			ERR_FAIL_V_MSG(-1, "Error calling method from 'first_custom': " + Variant::get_callable_error_text(p_callable, argptrs, 1, ce));
		}
		// Return as soon as possible
		if (result.operator bool()) {
			return i;
		}
	}

	return -1;
}

int Array::rfind_custom(const Callable& p_callable, int p_from = -1) const {
	if (_p->array.size() == 0) {
		return -1;
	}

	if (p_from < 0) {
		// Relative offset from the end
		p_from = _p->array.size() + p_from;
	}
	if (p_from < 0 || p_from >= _p->array.size()) {
		// Limit to array boundaries
		p_from = _p->array.size() - 1;
	}

	const Variant* argptrs[1];

	for (int i = p_from; i >= 0; i--) {
		argptrs[0] = &get(i);

		Variant result;
		Callable::CallError ce;

		p_callable.callp(argptrs, 1, result, ce);
		if (ce.error != Callable::CallError::CALL_OK) {
			ERR_FAIL_V_MSG(-1, "Error calling method from 'last_custom': " + Variant::get_callable_error_text(p_callable, argptrs, 1, ce));
		}
		// Return as soon as possible
		if (result.operator bool()) {
			return i;
		}
	}

	return -1;
}

int Array::count_custom(const Callable& p_callable) const {
	int count = 0;
	const Variant* argptrs[1];

	for (int i = 0; i < size(); i++) {
		argptrs[0] = &get(i);

		Variant result;
		Callable::CallError ce;

		p_callable.callp(argptrs, 1, result, ce);
		if (ce.error != Callable::CallError::CALL_OK) {
			ERR_FAIL_V_MSG(0, "Error calling method from 'count_custom': " + Variant::get_callable_error_text(p_callable, argptrs, 1, ce));
		}
		// Increase count if passes
		if (result.operator bool()) {
			count++;
		}
	}

	return count;
}
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Yes, one can simply create a loop to accomplish the same task. But it must be done every time, and could be solved by patching this gap in GDScript.

### Is there a reason why this should be core and not an add-on in the asset library?

This adds core Array functionality.

### PR DESCRIPTION
**Pull request by matt08-prog**
implemented the function Array::find_custom as discussed in [https://github.com/godotengine/godot-proposals/issues/6006](https://github.com/godotengine/godot-proposals/issues/6006).

This is my first time making a PR for the codebase so I'm sure I missed some things. Thanks in advance.

* Bugsquad edit, closes: https://github.com/godotengine/godot-proposals/issues/6889*

### CODE REVIEW

File: core/variant/array.cpp [hunk 1]
```diff
@@ -563,6 +563,32 @@ bool Array::any(const Callable &p_callable) const {
 	return false;
 }
 
+int Array::find_custom(const Callable &p_callable, const int from) const {
+	const Variant *argptrs[1];
+
+	ERR_FAIL_INDEX_V_MSG(from, size(), -1,
```
Comment (idbrii) [inline core/variant/array.cpp:569]: I think replace this check with an abort when `from` is negative like `find`. `find` and `rfind` don't error on the start position which prevents users from needing to check bounds before calling. If from is past the end, the loop won't execute.

**Review by idbrii — COMMENTED** [review]
Looks pretty good. Thanks for taking this on!

File: core/variant/array.cpp [hunk 2]
```diff
@@ -563,6 +563,32 @@ bool Array::any(const Callable &p_callable) const {
...
+		argptrs[0] = &get(i);
+
+		Variant result;
+		Callable::CallError ce;
+		p_callable.callp(argptrs, 1, result, ce);
+		if (ce.error != Callable::CallError::CALL_OK) {
+			ERR_FAIL_V_MSG(false, "Error calling method from 'find custom': " + Variant::get_callable_error_text(p_callable, argptrs, 1, ce));
+		}
+
+		if (result.operator bool()) {
```
[the remark this answers is not in the document]
Comment (matt08-prog) [reply core/variant/array.cpp:583]: The way this is handled in Array seems to be the norm for Godot. I implemented all of your other suggestions. Thank you.

**Comment by Maran23** [comment]
Looking at the proposal, the idea is to return an element (`Variant`) of the `Array`, not an `int`...
And as also written in the proposal, I'm still in favor of naming it `find_first`/`get_first`, to provide the information to the caller that he will get in fact the first element, so the sort order is important.

**Comment by AThousandShips** [comment]
Reviewers, see also:
* https://github.com/godotengine/godot/pull/73490
* https://github.com/godotengine/godot/pull/80208

**Comment by akien-mga** [comment]
Superseded by #95449. Thanks for the contribution!

**Comment by Radivarig** [comment]
@akien-mga This PR implements `Variant get_first` which was not covered by the PR it was superseded by #95449.

I was looking forward to having an array method that returns the Variant directly. Can this be reopened please?

**Pull request closed without merging**